pkg/csr/csr.go

```diff
@@ -0,0 +1,219 @@
+package csr
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"github.com/pkg/errors"
+	"net"
+	"reflect"
+	"strings"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=certificatesigningrequests/approval,verbs=update
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=certificatesigningrequests,verbs=get;list;watch
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=signers,verbs=Approve,resourceNames=kubernetes.io/kube-apiserver-client-kubelet;kubernetes.io/kubelet-serving
+
+var kubeletClientUsages = []certificatesv1.KeyUsage{
+	certificatesv1.UsageKeyEncipherment,
+	certificatesv1.UsageDigitalSignature,
+	certificatesv1.UsageClientAuth,
+}
+
+var kubeletServerUsages = []certificatesv1.KeyUsage{
+	certificatesv1.UsageKeyEncipherment,
+	certificatesv1.UsageDigitalSignature,
+	certificatesv1.UsageServerAuth,
+}
+
+const (
+	nodeGroup    = "system:nodes"
+	nodePrefix   = "system:node:"
+	systemPrefix = "system:authenticated"
+)
+
+var log = ctrl.Log.WithName("Config")
+
+// Config holds the information required to authorize and approve a CSR
+type Config struct {
+	k8sClientSet      *kubernetes.Clientset
+	instanceAddresses []string
+}
+
+// NewConfig returns a pointer to the Config
+func NewConfig(k8sclienSet *kubernetes.Clientset, instanceAddresses []string) *Config {
+	if len(instanceAddresses) == 0 || k8sclienSet == nil {
+	}
+	return &Config{k8sclienSet, instanceAddresses}
+}
+
+// Authorize returns true if the CSR request contents can be validated
+func (c *Config) Authorize(csr *certificatesv1.CertificateSigningRequest) (bool, error) {
+	parsedCSR, err := parseCSR(csr)
+	if err != nil {
+		return false, errors.Wrapf(err, "error parsing CSR %s", csr.Name)
+	}
+
+	dnsAddr := strings.TrimPrefix(parsedCSR.Subject.CommonName, nodePrefix)
+	if dnsAddr == "" {
+		return false, errors.Errorf("CSR subject name does not contain DNS address %s", csr.Name)
+	}
+
+	// check DNS of node requesting CSR matches instance IP/DNS address
+	var hasMatch bool
+	for _, instanceAdd := range c.instanceAddresses {
+		if areEqualAddresses(instanceAdd, dnsAddr) {
+			hasMatch = true
+			break
+		}
+	}
+	if !hasMatch {
+		return false, errors.Errorf("CSR is not from an authorized source %s", csr.Name)
+	}
+
+	if isNodeClientCert(csr, parsedCSR) {
+		// this is done before the instance becomes a node
+		return c.authorizeNodeClientCSR(dnsAddr)
+	} else {
+		return c.authorizeServerCSR(csr, parsedCSR)
+	}
+}
+
+// authorizeServerCSR authorizes a kubelet server CSR
+func (c *Config) authorizeServerCSR(req *certificatesv1.CertificateSigningRequest, csr *x509.CertificateRequest) (bool, error) {
```

Comment:
Server as in a master node?
We should never approve a master node CSR.

If this is not about master nodes then a different name and more descriptive comment should be used